tpchgen-cli/src/plan.rs

```diff
@@ -0,0 +1,349 @@
+    pub fn new(
+        table: &Table,
+        format: OutputFormat,
+        scale_factor: f64,
+        cli_part: i32,
+        cli_part_count: i32,
+    ) -> Self {
+        // parallel generation disabled if user specifies a part explicitly
+        if cli_part != -1 || cli_part_count != -1 {
+            return Self {
+                part_count: cli_part_count,
+                part_list: vec![cli_part],
+            };
+        }
+
+        // Note use part=1, part_count=1 to calculate the total row count
+        // for the table
+        //
+        // Avg row size is an estimate of the average row size in bytes from the first 100 rows
+        // of the table in tbl format
+        let (avg_row_size_bytes, row_count) = match table {
+            Table::Nation => (88, 1),
+            Table::Region => (77, 1),
+            Table::Part => (115, PartGenerator::calculate_row_count(scale_factor, 1, 1)),
+            Table::Supplier => (
+                140,
+                SupplierGenerator::calculate_row_count(scale_factor, 1, 1),
+            ),
+            Table::Partsupp => (
+                148,
+                PartSuppGenerator::calculate_row_count(scale_factor, 1, 1),
+            ),
+            Table::Customer => (
+                160,
+                CustomerGenerator::calculate_row_count(scale_factor, 1, 1),
+            ),
+            Table::Orders => (114, OrderGenerator::calculate_row_count(scale_factor, 1, 1)),
+            Table::Lineitem => {
+                // there are on average 4 line items per order.
+                // For example, in SF=10,
+                // * orders has 15,000,000 rows
+                // * lineitem has around 60,000,000 rows
+                let row_count = 4 * OrderGenerator::calculate_row_count(scale_factor, 1, 1);
+                (128, row_count)
+            }
+        };
+        // target chunks of about 16MB (use 15MB to ensure we don't exceed the target size)
+        let target_chunk_size_bytes = 15 * 1024 * 1024;
+        let mut num_parts = ((row_count * avg_row_size_bytes) / target_chunk_size_bytes) + 1;
+
+        // parquet files can have at most 32767 row groups so cap the number of parts at that number
+        if format == OutputFormat::Parquet {
+            num_parts = num_parts.min(32767);
+        }
+
+        // convert to i32
+        let num_parts = num_parts.try_into().unwrap();
+        // generating all the parts
+
+        Self {
+            part_count: num_parts,
+            part_list: (1..=num_parts).collect(),
+        }
+    }
+}
+
+/// Converts the `GenerationPlan` into an iterator of (num_parts, part_number)
+impl IntoIterator for GenerationPlan {
+    type Item = (i32, i32);
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.part_list
+            .into_iter()
+            .map(|part_number| (self.part_count, part_number))
```

Comment:
For anyone following along this is a bug, it should be

```suggestion
            .map(|part_number| (part_number, self.part_count))
```

I fixed it in
https://github.com/clflushopt/tpchgen-rs/pull/156